### ISSUE TITLE
[Advice] Group function for the vocabulary builder 

### ISSUE THREAD

**Issue by Lezcool** [koreader/koreader#9486]
Hi, I found vocabulary builder very helpful. But I'm learning 2 languages, so I hope there's a group function to separate them. 
Also compare to automatically adding words to vocabulary builder, I wish I can add word manually, because there's always mistouch.

**Comment by Lezcool** [comment]
I think I found a quick but unperfected solution for the 'manually adding word function', for those who also need this function.

Delete code below (900th line in my case) of readerdictionary.lua  (file path: `./koreader/frontend/apps/reader/modules`)
`self.ui:handleEvent(Event:new("WordLookedUp", word, book_title))`

and add the same code to readerhighlight.lua (same file path) under the 
``
local book_title = self.ui.doc_settings and self.ui.doc_settings:readSetting("doc_props").title or _("Dictionary lookup")
    if book_title == "" then -- no or empty metadata title
        if self.ui.document and self.ui.document.file then
            local directory, filename = util.splitFilePathName(self.ui.document.file) -- luacheck: no unused
            book_title = util.splitFileNameSuffix(filename)
        end
    end
    self.ui:handleEvent(Event:new("WordLookedUp", self.selected_text.text, book_title))
``

Then, I think the word will be added to vocabulary builder only when the word is highlighted.

**Comment by weijiuqiao** [comment]
Sorry for not replying more in time. It's better to just pin me if you want something with the vocab builder.

>But I hope there's a group function to group different type of words.

By grouping different types of words, do you mean a user can create any number of word types and then tag each word one? The plugin itself wouldn't know what type a given word is.

>Also compare to automatically adding words to vocabulary builder, I wish I can add word manually, because there's always mistouch and activate the dictionary, but don't want to add that word to vocabulary builder.

That would need a separate button. Also, one way is having to manually add the looked-up words as you suggested, and the other is to manually delete ones added by mistouch which is what it's now. Since I thought the former should happen more often than the latter, I chose the latter to save some manpower :)

**Comment by Lezcool** [comment]
Thanks for the reply. 
Yes, I mean user can make new groups, and add words to different groups they made. Or as you said, making new tags and tag different words is also nice. 
in order to to that, a separate button "add word" will be needed, after taping the button, users can choose the tag for the word. 
The main problem now is deleting word takes too many steps, and also usually I have to check the meaning of the word first and then decide if I need to add it to vocabulary builder.

**Comment by weijiuqiao** [comment]
You can use the "batch delete" function to delete unwanted words more easily.
Overall, I get what you want and why you want them, but to me it feels like many unnecessary steps. And also, adding groups means database modifications which I'm trying to avoid.
I see you've found a way to achieve one of the functions done, maybe we leave it like this for now and see if there are more people who want the same things?

**Comment by Lezcool** [comment]
Sure, thank you.

**Comment by weijiuqiao** [comment]
I find the "Close" button at the right corner of `dictquicklookup` kind of redundant as one can close it by tapping outside of the widget or using the "esc" button. Maybe we can customize it so that when a user is using the vocabulary builder plugin and choose to add words manually, this button is changed to something like `Add to vocal builder`?
Pinging @poire-z for the possibility :)

**Comment by poire-z** [comment]
In Dictionary settings, we have a `[ ] Large window` option, and tapping outside would be harder with it :)
`Add to vocabulary builder` would also be too wide.
_choose to add words manually_ : is this a current option ? Aren't all words added automatically?

Ideally, I think, adding more actions to this DictQuickLookup should better be done via a new widget that could be named PopupButton or PopupButtonTable, that would popup a single column stack of buttons: the <kbd>Highlight</kbd> would be <kbd>...</kbd>, and tapping on it would unfold some other buttons via this new widget. Then, plugins could register actions to be added in it (with may be some tools for re-ordering them, and may be assiging one to long-press on that <kbd>...</kbd>).
And ideally, this PopupButton widget could also be used for any top left/right button to popup a button list down under them.
(And that PopupButton width could adjust to either the largest button text, or some fixed width.)

**Comment by weijiuqiao** [comment]
>In Dictionary settings, we have a [ ] Large window option, and tapping outside would be harder with it :)

I failed to notice the `Large Window` option, but on second thought, adding a new button should be okay since we have plenty of room in this mode?


>`Add to vocabulary builder` would also be too wide.

Then `Add to vocabulary` or even `Add vocabulary`.


>choose to add words manually : is this a current option ? Aren't all words added automatically?

It's not a current option and yes all words are added automatically. It's what this FR is about though.

I played a little and since we had `tweak_buttons_func` already in place in `DictQuickLookup`, we can give it a default implementation in `vocabbuilder.koplugin`'s `main` depending on whether the conditions are met. Then it should all work. It is however altering what's inside `dictquicklookup` in the plugin so might have some coupling concerns.


>Ideally, I think, adding more actions to this DictQuickLookup should better be done via a new widget that could be named PopupButton or PopupButtonTable ...

I agree, maybe like how the long press of the keyboard keys' popup is implemented (haven't checked)? It's still not absolutely needed though even if we incorporate the manuel  `Add vocabulary`, so maybe not this time (if we are really adding this `Add vocabulary` button)?

**Comment by poire-z** [comment]
> adding a new button should be okay since we have plenty of room in this mode?
> since we had tweak_buttons_func already in place 
> It is however altering what's inside dictquicklookup in the plugin so might have some coupling concerns.

Why not - as long as by default (as your plugin is enabled by default), the DictQuickLookup window is not changed, as people may already be accustomed to the buttons layout - we should try to not mess with their habits :) until it's really necessary.
If your user enable your new option, you could even just add a wide button on top of the others. It might be ugly, dunno - but it gives you room for adding 1 or 2 more buttons later for new actions you can think of :)

> I agree, maybe like how the long press of the keyboard keys' popup is implemented (haven't checked)?

Yes, in that spirit. But its implementation in Keyboard relies on KeyPopup - so it's not reusable.

> so maybe not this time

Allright :) I think adding a PopupButton would be quite fun, but also tedious (its positionning, the refreshes, etc...)

**Comment by weijiuqiao** [comment]
>as long as by default (as your plugin is enabled by default), the DictQuickLookup window is not changed
If your user enable your new option

I'm thinking about changing the previous wording `Accept new words` (on by default) to `Auto add new words` so the buttons stay the same by default. This will however show the button for people who had disabled `Accept new words`. They'll have to disable the plugin altogether if they don't want to see the new button. It's still a bit intruding, should I just add a new option?

>you could even just add a wide button on top of the others

Then we can go full descriptive: `Add to vocabulary builder`

**Comment by poire-z** [comment]
> This will however show the button for people who had disabled Accept new words.

How do these people add words when this is disabled? Or is disabling this option is just for when you want to totally disable this plugin gathering words, when for example you're on a book in another language you don't want polluting your vocabulary?

> They'll have to disable the plugin altogether if they don't want to see the new button. It's still a bit intruding, should I just add a new option?

It feels like it should be an option, but I dunno your customers :)
May be a menu item with 3 states (with more clever/shorter English than mine:) : 
```
( ) Disable new words addition
(x) Add all lookup words
( ) Manually add lookup word with an added wide button
```

**Comment by weijiuqiao** [comment]
> Or is disabling this option is just for when you want to totally disable this plugin gathering words

Yes. Like you can stop having new words while still work on the old ones in the plugin.

On second thought, for those who have just temporarily disabled getting new words, it wouldn't bother them too much to have a manual button. One might want to occasionally add favored ones! For those who disable the option because they don't use the plugin at all, maybe a new button is a remainder that they can disable the plugin altogether.

So to keep things simple, I think I'll just repurpose the old `Accept new words` button :)

**Comment by weijiuqiao** [comment]
It turns out we can “regroup” the words by assigning them to a different book. I’m considering making it re-assignable and give user ability to add virtual books. The assignment UI needs some more thinking, though.

**Comment by Lezcool** [comment]
Thank you~ And please make an update once finishing adding the features, since I’m currently avoiding updates in case overwrite my own modifications.
Sent with Spark

**Comment by weijiuqiao** [comment]
The functionality has been merged and is available in the latest nightly build. 
>since I’m currently avoiding updates in case overwrite my own modifications.

In case you didn't know, if you use User patches, modifications can survive OTA updates: https://github.com/koreader/koreader/wiki/User-patches

**Comment by Lezcool** [comment]
Cool. Thanks a lot.
